### PR TITLE
Preserve original publish dates when backfilling with --original-time

### LINKED ISSUE #212: Feature Request: Implement --original-time so backfilled records sort by their original publish date
### What

Implement the `--original-time` flag on `wp atmosphere backfill` so backfilled records sort by each post's **original publish date** in Bluesky and standard.site readers, instead of by the time the backfill ran.

The flag already exists but is a no-op: it's parsed and then warns and does nothing (`includes/cli/class-backfill-command.php:160-173`).

```php
if ( $original_time ) {
    \WP_CLI::warning(
        \__( '--original-time is not yet implemented and has no effect; records use the default timestamp behavior. Support is planned for a future release.', 'atmosphere' )
    );
}
```

### Why

When you backfill an existing site, the record's `createdAt` is already set to the original post date (`Publisher::publish_single()`, ~line 316-317). But the record **key** (rkey) is a TID minted from the *current* time, and the rkey — not `createdAt` — is what determines sort order in feeds and readers. So a backfill of an old catalogue makes every historical post appear brand-new at the top of the timeline / reader, which is exactly the "flooding" problem the `--original-time` flag was reserved to solve.

This matters most for archival use: seeding a long back catalogue as standard.site documents (see the document-only filter in #211) so the content reads in chronological order in tools like Standard Reader, rather than all bunched at "now."

### How

The hard part is TID minting. `TID::generate()` (`includes/transformer/class-tid.php:73`) derives the TID from current `microtime()` and enforces a **persisted monotonic floor** (`OPTION_LAST_TS`) via an atomic compare-and-set — the floor must never regress, so concurrent normal publishes always get strictly-increasing TIDs. A historical TID is, by definition, *below* that floor, so you cannot route it through the existing path without either regressing the floor (breaking the invariant for live publishes) or having the floor bump it back up to "now" (defeating the point).

A workable approach for the implementer to evaluate:

- Add a historical-TID path that derives a TID from `post_date_gmt` **without** reading or advancing `OPTION_LAST_TS` — historical TIDs live below the live floor and simply don't participate in the monotonic guarantee (which only needs to hold for concurrent *live* minting).
- Handle collisions among backfilled posts (posts sharing a second/microsecond, bulk imports) using the existing 10-bit `clock_id` + per-process counter, or a deterministic per-post disambiguator, so two historical posts never collide on the same rkey.
- Wire the flag through so it reaches TID minting. Note both the Bluesky post rkey **and** the document rkey come from `get_rkey()` → `TID::generate()`, and `get_rkey()` **persists** the TID as post meta (`Post::META_TID` / `Document::META_TID`) on first call — so the original-time decision has to happen at/before that reservation, and must apply uniformly to both records (and to the document-only path from #211).

Scope notes / non-goals to confirm:

- rkeys are immutable identity once a record is created, so this can only affect records at **initial create** — it can't retroactively re-sort already-synced posts. Re-minting existing records is out of scope (would mean delete + republish, orphaning any engagement).
- Decide what happens to the reserved-but-unpublished TID case (a post that reserved a `now` TID from a prior failed attempt, then gets backfilled with `--original-time`).

Related: #91 (added the backfill command), #211 (document-only publishing filter — the primary use case for chronological archival seeding).

### PR DESCRIPTION
**Pull request by jeherve**
Fixes #212

## Proposed changes:

`wp atmosphere backfill --original-time` used to be a parsed no-op — it warned and did nothing. Now it mints each backfilled record's key (the AT Protocol TID that determines sort order) from the post's **original publish date** instead of the moment the backfill ran.

Backfilled records already carried the correct `createdAt`, but feeds and readers (Bluesky, Standard Reader) sort by the record **key**, not `createdAt`. So seeding an old catalogue dumped every historical post at the top of the timeline as if brand-new. With `--original-time`, an archive reads back in chronological order — the main point of seeding a back catalogue as standard.site documents.

**Why the TID path needed care.** `TID::generate()` enforces a persisted monotonic floor that must never regress, so live concurrent publishes always get strictly increasing keys. A past-dated TID sits *below* that floor, so it can't go through the normal path without either regressing the floor or being bumped back up to "now". The new `TID::generate_for_time()` mints from a given past time **without** reading or advancing the floor — historical keys simply don't participate in the monotonic guarantee, which only needs to hold among live mints.

**Collision handling.** WordPress `post_date_gmt` is second-precision, so two posts sharing a second would otherwise encode the identical key and the PDS would reject the duplicate create. The post ID is folded into the unused sub-second slot (`micros = seconds * 1e6 + post_id % 1e6`) — deterministic, and it orders same-second posts by creation order. Teaser-thread replies are minted historically too, sequenced right after their root.

**Deliberately scoped to initial creation.** Record keys are immutable identity, so this only affects a post's *first* publish. Already-synced posts — and posts that reserved a key from a prior failed attempt — keep their existing key; `--original-time` never re-mints. Re-sorting already-published records is out of scope (it would mean delete + republish, orphaning any engagement). Only `Publisher::publish_post()` was touched; the update/delete paths reuse stored keys unchanged.

Developer- and CLI-facing only — no UI changes.

### Other information:

- [x] Have you written new tests for your changes, if applicable?

New coverage: `TID::generate_for_time()` / `decode()` (floor left untouched, sort order, sub-second disambiguation, bad-date fallback), transformer historical minting with idempotency and reserved-key preservation, `Publisher::publish_post()` reserving historical keys for both the Bluesky post and the document, and the CLI flag end to end. Full suite: 1065 tests passing; `composer lint` clean.

## Testing instructions:

1. Connect the plugin to an AT Protocol account.
2. Create a few published posts with backdated dates, left unsynced:
   ```
   wp post create --post_status=publish --post_date='2019-05-04 09:00:00' --post_title='Old post'
   ```
3. Preview: `wp atmosphere backfill --original-time --dry-run`.
4. Run it: `wp atmosphere backfill --original-time`. Note the one-line "original time only affects posts published for the first time" message.
5. In the AT Protocol repo, confirm each record's rkey encodes the original date — the documents sort chronologically in a reader (e.g. Standard Reader) instead of bunching at "now". For contrast, backfill another post *without* the flag and confirm its rkey encodes ~now.
6. Re-run `wp atmosphere backfill --original-time --force` on an already-synced post and confirm its AT-URIs are unchanged (keys are immutable; the flag is a no-op there).

### Changelog entry

A changelog entry is already committed on the branch (`.github/changelog/add-backfill-original-time`), so the auto-create box below is intentionally left unchecked.

### CODE REVIEW

File: includes/transformer/class-tid.php [hunk 1]
```diff
@@ -167,22 +167,82 @@ public static function generate(): string {
...
+		if ( $unix_seconds <= 0 ) {
+			// Unparseable / zero date (e.g. `0000-00-00`): mint a live TID
+			// rather than a garbage epoch-1970 rkey.
+			return self::generate();
+		}
+
+		$micros = $unix_seconds * 1_000_000 + ( $disambiguator % 1_000_000 );
+
+		return self::encode( ( $micros << 10 ) | self::clock_id() );
+	}
```
[the remark this answers is not in the document]
Comment (jeherve) [reply includes/transformer/class-tid.php:259]: Addressed in c879273.

- **Negative disambiguator** — good catch. `generate_for_time()` now reduces the disambiguator with a sign-safe wrap into `[0, 1,000,000)`, so a negative value can no longer push the key before the target second. Added `test_generate_for_time_negative_disambiguator_stays_in_second`.
- **Same-second collision** — the docblock overstated the guarantee, so I softened it to "very unlikely to collide" and documented the exact residual: two roots coincide only if their IDs are congruent modulo 100,000 within the same second. (Since 9d05ac9 reserves the low decimal digit of the slot for the reply sequence, the post ID disambiguates modulo 100,000.) In the bulk/WXR-import case this targets, IDs run consecutively, so that needs 100,000 posts in one second — negligible.

File: includes/transformer/class-tid.php [hunk 2]
```diff
@@ -167,22 +167,82 @@ public static function generate(): string {
...
+	/**
+	 * Lazily seed and return the per-process 10-bit clock identifier.
+	 *
+	 * `random_int` throws only on a system without a usable CSPRNG
+	 * (essentially never on a working PHP install); `wp_rand` is the
+	 * non-cryptographic fallback. Shared by {@see self::generate()} and
+	 * {@see self::generate_for_time()} so both mint from the same clock.
+	 *
+	 * @return int
+	 */
```
[the remark this answers is not in the document]
Comment (jeherve) [reply includes/transformer/class-tid.php:186]: Clarified in c879273. The comment was accurate — the clock ID is a single per-process constant shared by `generate()` and `generate_for_time()` — but I reworded it to head off exactly the future-refactor confusion you flagged: the clock bits identify the worker and do **not** disambiguate two mints landing on the same microsecond. That job belongs to the sub-second slot for historical mints and to the monotonic floor for live ones; the clock is never derived from the disambiguator.

File: includes/transformer/class-base.php [hunk 3]
```diff
@@ -52,6 +52,44 @@ public function __construct( mixed $object ) { // phpcs:ignore Universal.NamingC
...
+	/**
+	 * Mint a historical TID from the post's original publish date.
+	 *
+	 * The post ID fills the sub-second slot so two posts sharing a
+	 * second don't collide (see {@see TID::generate_for_time()});
+	 * `$sequence` offsets a thread reply after its root.
+	 *
+	 * @param int $sequence Offset within the post's records (0 = root/doc).
+	 * @return string
+	 */
```
[the remark this answers is not in the document]
Comment (jeherve) [reply includes/transformer/class-base.php:105]: Softened in c879273. Correct — the disjoint-range scheme (9d05ac9) reserves a low decimal digit for the reply sequence, so post IDs disambiguate modulo 100,000, and two roots collide only if their IDs are congruent modulo 100,000 in the same publish second. The docblock now says "very unlikely to collide" and spells that residual out, noting it is negligible because import IDs run consecutively.

File: includes/transformer/class-post.php [hunk 4]
```diff
@@ -761,13 +761,29 @@ public function get_rkey(): string {
...
+	 * on the reply is dated just after the root within the same second;
+	 * otherwise a fresh live TID is used. Not persisted here — replies
+	 * are tracked in `META_THREAD_RECORDS` by Publisher.
+	 *
+	 * @param int $index Reply index within the thread (>= 1).
+	 * @return string
+	 */
+	public function mint_reply_rkey( int $index ): string {
+		return $this->original_time ? $this->historical_rkey( $index ) : TID::generate();
+	}
```
[the remark this answers is not in the document]
Comment (jeherve) [reply includes/transformer/class-post.php:785]: Already fixed in 9d05ac9 (landed after this review). `historical_rkey()` no longer sums `post_id + $sequence`; it packs them into disjoint ranges of the sub-second slot — `(ID % 100,000) * 10 + $sequence` — so the reply sequence lives in a reserved low decimal digit and can never wrap the root's slot back to 0. The teaser thread is capped at 5 records, so one digit is ample headroom; reply keys always sort strictly after the root and can no longer coincide with a neighboring post's root. Covered by a regression test in that commit.

File: includes/transformer/class-tid.php [hunk 5]
```diff
@@ -167,22 +167,98 @@ public static function generate(): string {
...
+	public static function decode( string $tid ): int {
+		$value = 0;
+		$len   = \strlen( $tid );
+
+		for ( $i = 0; $i < $len; $i++ ) {
+			$value = ( $value << 5 ) | (int) \strpos( self::CHARSET, $tid[ $i ] );
+		}
+
+		return $value >> 10;
 	}
```
[the remark this answers is not in the document]
Comment (jeherve) [reply includes/transformer/class-tid.php:292]: Fixed in 298e76a. `decode()` now guards on `is_valid()` (length + charset) and returns 0 for malformed input, so a stray non-charset character or wrong-length string no longer decodes into a plausible-but-wrong timestamp. 0 is a safe sentinel here — a real TID never decodes to 0 (the historical path falls back to a live TID for epoch-0 dates, and a live TID is minted from the current time). Added `test_decode_returns_zero_for_invalid_tid`.

File: tests/phpunit/tests/class-test-publisher.php [hunk 6]
```diff
@@ -2724,4 +2725,80 @@ public function test_publish_skips_publication_sync_when_in_sync() {
...
+					$results[] = array(
+						'uri' => 'at://did:plc:test123/' . ( $write['collection'] ?? 'x' ) . '/' . ( $write['rkey'] ?? '' ),
+						'cid' => 'bafyreib' . \substr( \md5( (string) ( $write['rkey'] ?? '' ) ), 0, 20 ),
+					);
+				}
+				return array( 'results' => $results );
+			},
+			10,
+			2
+		);
```
Comment (pfefferle) [inline tests/phpunit/tests/class-test-publisher.php:3284]: The existing `mock_response()` helper already builds this response shape.

```suggestion
		\add_filter(
			'atmosphere_pre_apply_writes',
			fn( $short_circuit, array $writes ) => $this->mock_response( $writes ),
			10,
			2
		);
```

File: tests/phpunit/tests/class-test-publisher.php [hunk 7]
```diff
@@ -2724,4 +2725,80 @@ public function test_publish_skips_publication_sync_when_in_sync() {
...
```
Comment (pfefferle) [inline tests/phpunit/tests/class-test-publisher.php:3314]: Same here, reuse `mock_response()`.

```suggestion
		\add_filter(
			'atmosphere_pre_apply_writes',
			fn( $short_circuit, array $writes ) => $this->mock_response( $writes ),
			10,
			2
		);
```

**Review by pfefferle — COMMENTED** [review]
Had a closer look at the historical rkey part. A few things, first one I'd like your take on.

The `(ID % 100000) * 10` disambiguator can collide. You call it out in the docblock, but I'm not sure "negligible" holds on big sites. Two roots collide when their IDs are congruent mod 100,000 and land in the same second. On a site with 100k+ posts a WXR import puts lots of posts in the same second, which is exactly the case this targets, so IDs 100,000 apart colliding isn't that rare. And it fails quietly: the second `#create` comes back as "record already exists" and that post just drops out of the backfill. Might need a wider disambiguator.

`generate_for_time()` never advances `OPTION_LAST_TS`. The docblock says a historical TID is by definition below the live floor, which holds for a normal old post. But a `publish` post can carry a future `post_date_gmt`, and then it mints above the floor without advancing it, so a concurrent live publish could regress against it. Edge case, but maybe clamp to now when the date is in the future.

`TID::decode()` has no production caller, only the tests use it. That means the base-32 inverse ships and has to track `generate_for_time`'s bit layout while nothing runs it. Could be a test helper, or round-trip against `generate_for_time` output instead.

Two small suggestions inline on the tests too. Not blocking on any of these.

File: tests/phpunit/tests/class-test-publisher.php (hunk 6 above)
Comment (jeherve) [reply tests/phpunit/tests/class-test-publisher.php:3284]: Applied in 1facde5 — both switched to `fn( $short_circuit, array $writes ) => $this->mock_response( $writes )`. Thanks!

File: tests/phpunit/tests/class-test-publisher.php (hunk 7 above)
Comment (jeherve) [reply tests/phpunit/tests/class-test-publisher.php:3314]: Applied in 1facde5 — both switched to `fn( $short_circuit, array $writes ) => $this->mock_response( $writes )`. Thanks!

**Comment by jeherve** [comment]
Thanks @pfefferle — all addressed.

**1. Disambiguator width (`12cac98`).** You're right that 100,000 slots isn't enough for a same-second bulk import — a date-only migration is the obvious case, where every post that day lands on `YYYY-MM-DD 00:00:00`. I widened it by packing the next slice of the ID into the TID's **10 clock-id bits**, which were previously just a random per-process constant that disambiguated nothing *within* a run. That lifts the effective per-second space from 100,000 to **~102.4 million** (100,000 × 1,024): two roots now collide only if their IDs are congruent modulo 102,400,000 in the same second — beyond the ID range of any realistic site, so same-second imports stop dropping posts. New tests cover it end-to-end (two posts with IDs congruent mod 100,000 → distinct rkeys) and at the unit level. On "fails quietly": the CLI does surface a failed `#create` as `Failed to publish post N` with a non-zero exit, but the wider space means it shouldn't come up. (If 102M ever feels tight we could spill into adjacent seconds once a second is exhausted, trading exact-second ordering for capacity — happy to if you want it.)

**2. Future dates (`12cac98`).** Added the clamp — `generate_for_time()` now caps `unix_seconds` at `now`, so a `publish` post dated in the future can't mint above the live floor and sort ahead of genuinely-later records.

**3. `decode()` (no production caller).** Fair — only the suite runs it today. I kept it as the natural inverse of the (public) TID encoding and a legitimate verify/debug utility ("what time does this rkey encode?"); the layout it tracks is a fixed, trivial transform (drop the low 10 clock bits + base-32 inverse) that can't silently drift, and the round-trip tests would catch it if it did. If you'd rather it not ship, I'm glad to move it into a test-only helper or switch the assertions to round-trip against `generate_for_time` output — just say the word.

Inline test suggestions applied in `1facde5`.

**Review by pfefferle — DISMISSED** [review]
thanks @jeherve, this all looks good now.

the clock-bit packing is the right fix. those 10 bits were just a random per-process constant before, so reusing them for the next slice of the id gets you ~102M of effective per-second space for free, well beyond any real site's id range. same-second imports stop dropping posts.

on `decode()`: keep it as is. it is a trivial fixed inverse of the encoding, the round-trip tests catch any drift, and it is a legit verify/debug helper. no need to hide it behind the test suite.

one thing before you merge: the branch is behind trunk, mind updating it?

**Comment by jeherve** [comment]
I'm afraid I'll need a final approval after the rebase. 
Thank you!

**Review by pfefferle — APPROVED** [review]
